lib/client.js

```diff
@@ -242,12 +242,20 @@ Client.prototype.close = function (done) {
 
   function finish () {
     if (!that.disconnected && that.will) {
-      that.broker.publish(that.will, that, function (err) {
+      that.broker.authorizePublish(that, that.will, function (err) {
         if (!err) {
-          that.broker.persistence.delWill({
-            id: that.id,
-            brokerId: that.broker.id
-          }, nop)
+          that.broker.publish(that.will, that, done)
+        } else {
+          done()
+        }
+
+        function done (err) {
+          if (!err) {
+            that.broker.persistence.delWill({
+              id: that.id,
+              brokerId: that.broker.id
+            }, nop)
+          }
         }
       })
       that.will = null // this function might be called twice
```

Comment:
Huhm! `authorizePublish` is async function. Server will crash when `that.broker.publish(that.will, that, done)` called because `that.will` is `null`